GetStream/stream-chat-android
# [4381] `MessageComposerView` video previews

### 🎯 Goal

closes #4381

### 🛠 Implementation details

- Renamed `ImageAttachmentPreviewFactory`  to `MediaAttachmentPreviewFactory` and gave it the ability to preview video attachments as well as image attachments
- Propagated `MessageComposerViewStyle` preview `ViewHolders` used for displaying attachments inside  `MessageComposerView` 

### 🎨 UI Changes

| Before | After |
| --- | --- |
| ![before](https://user-images.githubusercontent.com/37080097/200920239-abf43e02-9c35-4abd-a622-d33c316b2c3d.png) | ![after](https://user-images.githubusercontent.com/37080097/200920225-6447323c-3abf-43c7-9f13-8eb6ed564a1e.png) |

### 🧪 Testing

### Test A: Functionality Test

1. Launch the XML sample app
2. Enter a channel
3. Try sending various attachments such as video, image and file attachments

Expected: Video attachments should now be rendered as a thumbnail with an icon overlaid above them, others should be rendered the same as before.


### Test B: Functionality Test

1. Repeat Test A but this time try to add 10 attachments, some should be video, some image
2. Scroll from the beginning to the a couple of times 

Expected: The play button should appear only over videos

### Test C: Pressure Test

We've changed the point in which we initialize the adapter inside `DefaultMessageComposerCenterContent` so as a precautionary measure we'll do a lifecycle test.

 1. Launch the XML sample app
 3. Enter a channel and pick an attachment
 4. Don't send the attachment
 5. Try to put the app in the background and come back to the foreground 
 6. test any other lifecycle event you might think necessary
 
 Expected result: No crash, everything works as usual

### Test D: Customization Test

<details>

<summary>Patch that applies customization options</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml b/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml
--- a/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml	(revision 02cd8d1ca75516229541283d6ff74eb421342817)
+++ b/stream-chat-android-ui-components-sample/src/main/res/layout/fragment_chat.xml	(date 1668020004250)
@@ -51,6 +51,15 @@
         app:layout_constraintEnd_toEndOf="parent"
         app:layout_constraintStart_toStartOf="parent"
         app:layout_constraintTop_toBottomOf="@+id/messageListView"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconDrawable="@drawable/stream_ui_ic_reaction_thumbs_up"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconDrawableTint="@color/green"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconBackgroundColor="@color/red"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconElevation="12dp"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconDrawablePaddingTop="6dp"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconDrawablePaddingBottom="6dp"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconDrawablePaddingStart="6dp"
+        app:streamUiMessageComposerMessageInputVideoAttachmentIconDrawablePaddingEnd="6dp"
+
         />
 
 </androidx.constraintlayout.widget.ConstraintLayout>
Index: stream-chat-android-ui-components/src/main/res/values/styles.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components/src/main/res/values/styles.xml b/stream-chat-android-ui-components/src/main/res/values/styles.xml
--- a/stream-chat-android-ui-components/src/main/res/values/styles.xml	(revision 02cd8d1ca75516229541283d6ff74eb421342817)
+++ b/stream-chat-android-ui-components/src/main/res/values/styles.xml	(date 1668019931283)
@@ -1034,6 +1034,7 @@
         <item name="streamUiMessageComposerScrollbarFadingEnabled">false</item>
         <item name="streamUiMessageComposerMessageInputMaxLines">7</item>
         <item name="streamUiMessageComposerMessageInputCannotSendHintText">@string/stream_ui_message_composer_hint_cannot_send_message</item>
+
         <item name="streamUiMessageComposerMessageInputVideoAttachmentIconDrawable">@drawable/stream_ui_ic_play</item>
         <item name="streamUiMessageComposerMessageInputVideoAttachmentIconBackgroundColor">@color/stream_ui_literal_white</item>
         <item name="streamUiMessageComposerMessageInputVideoAttachmentIconElevation">3dp</item>
```

</details>

 1. Apply the patch above
 7. Launch the XML sample app
 8. Change customization options and verify that they are deployed correctly


### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

![gif](https://media1.giphy.com/media/DXMlAq43aP0Xe/giphy.gif?cid=ecf05e47ot77lbizdn97v0iqlk6uuhos9xnat8lo9lt6o6ui&rid=giphy.gif&ct=g)
